File: app/src/main/java/fr/legrand/oss117soundboard/data/repository/ContentRepositoryImpl.java
package fr.legrand.oss117soundboard.data.repository;

import android.content.Context;

import java.util.Collections;
import java.util.List;
import java.util.concurrent.ThreadLocalRandom;

import javax.inject.Inject;
import javax.inject.Singleton;

import fr.legrand.oss117soundboard.R;
import fr.legrand.oss117soundboard.data.entity.Reply;
import fr.legrand.oss117soundboard.data.manager.db.DatabaseManager;
import fr.legrand.oss117soundboard.data.manager.file.FileManager;
import fr.legrand.oss117soundboard.data.manager.sharedpref.SharedPrefManager;
import io.reactivex.Completable;
import io.reactivex.Observable;

/**
 * Created by Benjamin on 30/09/2017.
 */
@Singleton
public class ContentRepositoryImpl implements ContentRepository {

    private DatabaseManager databaseManager;
    private SharedPrefManager sharedPrefManager;
    private FileManager fileManager;
    private Context context;

    @Inject
    public ContentRepositoryImpl(DatabaseManager databaseManager, SharedPrefManager sharedPrefManager, FileManager fileManager, Context context) {
        this.databaseManager = databaseManager;
        this.sharedPrefManager = sharedPrefManager;
        this.fileManager = fileManager;
        this.context = context;
    }

    @Override
    public Observable<List<Reply>> getReplyWithSearch(String search, boolean fromFavorite) {
        return Observable.defer(() -> {
                    List<Reply> replyList = databaseManager.getReplyWithSearch(search, fromFavorite);
                    if (sharedPrefManager.getReplySort() == null || sharedPrefManager.getReplySort().equals(context.getString(R.string.alphabetical_order))) {
                        Collections.sort(replyList, (r1, r2) -> r1.getName().compareTo(r2.getName()));
                    } else if (sharedPrefManager.getReplySort().equals(context.getString(R.string.random_order))) {
                        Collections.shuffle(replyList);
                    }
                    return Observable.just(replyList);
                }
        );
    }

    @Override
    public Completable initAllReply() {
        return Completable.defer(() ->
                Completable.fromCallable(() ->
                        databaseManager.getAllReply())
                        .onErrorResumeNext(throwable -> {
                            List<Reply> replyList = fileManager.buildReplyListFromResource();
                            databaseManager.saveReplyList(replyList);
                            return Completable.complete();
                        })
        );
    }

    @Override
    public Observable<Reply> updateFavoriteReply(int replyId, boolean addToFavorite) {
        return Observable.defer(() -> Observable.just(databaseManager.updateFavoriteReply(replyId, addToFavorite)));
    }

    @Override
    public Observable<List<Reply>> getAllReply(boolean fromFavorite) {
        return Observable.defer(() -> {
            List<Reply> replyList = databaseManager.getAllReply(fromFavorite);
            if (sharedPrefManager.getReplySort() == null || sharedPrefManager.getReplySort().equals(context.getString(R.string.alphabetical_order))) {
                Collections.sort(replyList, (r1, r2) -> r1.getName().compareTo(r2.getName()));
            } else if (sharedPrefManager.getReplySort().equals(context.getString(R.string.random_order))) {
                Collections.shuffle(replyList);
            }
            return Observable.just(replyList);
        });
    }

    @Override
    public Completable updateMultiListenParameter(boolean multiListen) {
        sharedPrefManager.setMultiListenEnabled(multiListen);
        return Completable.complete();
    }

    @Override
    public boolean isMultiListenEnabled() {
        return sharedPrefManager.isMultiListenEnabled();
    }

    @Override
    public Observable<Reply> getMostListenedReply() {
        return Observable.fromCallable(() -> databaseManager.getMostListenedReply());
    }

    @Override
    public Completable incrementReplyListenCount(int replyId) {
        return Completable.defer(() -> {
            databaseManager.incrementReplyListenCount(replyId);
            return Completable.complete();
        });
    }

    @Override
    public void increaseTotalReplyTime(long duration) {
        sharedPrefManager.increaseTotalReplyTime(duration);
    }

    @Override
    public Observable<Long> getTotalReplyTime() {
        return Observable.fromCallable(() -> sharedPrefManager.getTotalReplyTime());
    }

    @Override
    public Observable<Integer> getRandomReplyIdToListen() {
        return Observable.defer(() -> {
            List<Reply> replyList = databaseManager.getAllReply();
            int randomNumber = ThreadLocalRandom.current().nextInt(0, replyList.size() + 1);
            int replyId = replyList.get(randomNumber).getId();
            databaseManager.incrementReplyListenCount(replyId);
            return Observable.just(replyId);
        });
    }

    @Override
    public Completable updateReplySort(String replySort) {
        return Completable.defer(() -> {
            sharedPrefManager.setReplySort(replySort);
            return Completable.complete();
        });
    }

    @Override
    public Observable<String> getReplySort() {
        return Observable.fromCallable(() -> sharedPrefManager.getReplySort());
    }
}

File: app/src/main/java/fr/legrand/oss117soundboard/data/repository/ContentRepositoryImpl.kt
package fr.legrand.oss117soundboard.data.repository

import android.content.Context
import fr.legrand.oss117soundboard.R
import fr.legrand.oss117soundboard.data.entity.Reply
import fr.legrand.oss117soundboard.data.manager.db.DatabaseManager
import fr.legrand.oss117soundboard.data.manager.file.FileManager
import fr.legrand.oss117soundboard.data.manager.sharedpref.SharedPrefManager
import io.reactivex.Completable
import io.reactivex.Observable
import java.util.concurrent.ThreadLocalRandom
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Benjamin on 30/09/2017.
 */
@Singleton
class ContentRepositoryImpl @Inject constructor(
        private val databaseManager: DatabaseManager,
        private val sharedPrefManager: SharedPrefManager,
        private val fileManager: FileManager,
        private val context: Context
) : ContentRepository {

    override fun getReplyWithSearch(search: String, fromFavorite: Boolean): Observable<List<Reply>> {
        return Observable.defer {
            val replyList = databaseManager.getReplyWithSearch(search, fromFavorite)
            when {
                sharedPrefManager.getReplySort() == context.getString(R.string.alphabetical_order) ->
                    replyList.sortedBy { it.name }
                sharedPrefManager.getReplySort() == context.getString(R.string.movie_order) ->
                    replyList.sortedBy { it.timestamp }
                sharedPrefManager.getReplySort() == context.getString(R.string.random_order) ->
                    replyList.shuffled()
            }
            Observable.just(replyList)
        }
    }

    override fun initAllReply(): Completable {
        return Completable.defer {
            if (databaseManager.getAllReply().isEmpty()) {
                val replyList = fileManager.buildReplyListFromResource()
                databaseManager.saveReplyList(replyList)
            }
            Completable.complete()
        }
    }

    override fun updateFavoriteReply(replyId: Int, addToFavorite: Boolean): Observable<Reply> {
        return Observable.defer { Observable.just(databaseManager.updateFavoriteReply(replyId, addToFavorite)) }
    }

    override fun getAllReply(fromFavorite: Boolean): Observable<List<Reply>> {
        return Observable.defer {
            val replyList = databaseManager.getAllReply(fromFavorite)
            when {
                sharedPrefManager.getReplySort() == context.getString(R.string.alphabetical_order) ->
                    replyList.sortedBy { it.name }
                sharedPrefManager.getReplySort() == context.getString(R.string.movie_order) ->
                    replyList.sortedBy { it.timestamp }
                sharedPrefManager.getReplySort() == context.getString(R.string.random_order) ->
                    replyList.shuffled()
            }
            Observable.just(replyList)
        }
    }

    override fun updateMultiListenParameter(multiListen: Boolean): Completable {
        sharedPrefManager.setMultiListenEnabled(multiListen)
        return Completable.complete()
    }

    override fun multiListenEnabled(): Observable<Boolean> {
        return Observable.fromCallable { sharedPrefManager.isMultiListenEnabled() }
    }

    override fun getMostListenedReply(): Observable<Reply> {
        return Observable.fromCallable { databaseManager.getMostListenedReply() }
    }

    override fun incrementReplyListenCount(replyId: Int): Completable {
        return Completable.defer {
            databaseManager.incrementReplyListenCount(replyId)
            Completable.complete()
        }
    }

    override fun increaseTotalReplyTime(duration: Long) {
        sharedPrefManager.increaseTotalReplyTime(duration)
    }

    override fun getTotalReplyTime(): Observable<Long> {
        return Observable.fromCallable { sharedPrefManager.getTotalReplyTime() }
    }

    override fun retrieveRandomReplyIdToListen(): Observable<Int> {
        return Observable.defer {
            val replyList = databaseManager.getAllReply()
            val randomNumber = ThreadLocalRandom.current().nextInt(0, replyList.size + 1)
            val replyId = replyList[randomNumber].id
            databaseManager.incrementReplyListenCount(replyId)
            Observable.just(replyId)
        }
    }

    override fun updateReplySort(replySort: String): Completable {
        return Completable.defer {
            sharedPrefManager.setReplySort(replySort)
            Completable.complete()
        }
    }

    override fun getReplySort(): Observable<String> {
        return Observable.fromCallable { sharedPrefManager.getReplySort() }
    }
}